Dit 93 rotate secrets

chris01b opened the pull request:
# Rotate Secrets & Secrets Manager Lambda Extension

## Summary

This pull request introduces a robust, automated secret rotation mechanism for the application's `FLASK_SECRET_KEY` and centralizes all secret retrieval logic into a new, optimized shared module. This enhances security by regularly rotating a critical credential and improves performance and cost-effectiveness by leveraging the AWS Parameters and Secrets Lambda Extension.

Note: When deploying wearable-data-retrieval, you need to make the secrets extension available by using code from deploy-staging/prod.sh

## Changes

#### 1. Automatic Flask Secret Key Rotation

A new Lambda function (`functions/secret_rotator`) has been implemented to handle the automatic rotation of the `FLASK_SECRET_KEY` in AWS Secrets Manager.

*   **Robust Test-Driven Rotation**: The rotation process includes a vital `test` step that verifies the application has successfully loaded the new (pending) secret before it's promoted to `AWSCURRENT`.
*   **Health Check Endpoint**: A new `/healthz` endpoint was added to the Flask. It returns the `VersionId` of the currently loaded Flask secret key, allowing the rotator to confirm the update was successful.
*   **Deployment Orchestration**: The `deploy-staging.sh` and `deploy-prod.sh` scripts have been significantly updated to manage the deployment of the rotator function. They now build and push the rotator's Docker image, deploy it with Zappa, and then imperatively apply the necessary IAM policies and environment variables (`APP_URL`, `APP_LAMBDA_FUNCTION_NAME`) for the rotator to function correctly.

#### 2. Centralized and Optimized Secret Retrieval

A new shared module, `shared/secrets.py`, has been created to provide a single, consistent interface for retrieving secrets.

*   **Lambda Extension-Aware**: The `get_secret` function first attempts to fetch secrets from the AWS Parameters and Secrets Lambda Extension. This is the preferred method in a Lambda environment for high performance and reduced cost. If the extension is unavailable (e.g., local development), it transparently falls back to a standard `boto3` API call.
*   **Code Refactoring**: The following components have been refactored to use the new centralized `get_secret` function, inheriting its performance benefits:
    *   `run.py` (application startup)
    *   `shared/tokens_manager.py`
    *   `backend/utils/lambda_credentials_manager.py`
    *   `functions/wearable_data_retrieval/lambda_function.py`
    *   `rds_stopper.py`

#### 3. Build & CI/CD Improvements

*   **Lambda Extension in Docker**: The deployment scripts now download the official AWS Lambda extension and the Dockerfiles copy it into the production images, ensuring it's available at runtime.
*   **CI Build Fix**: The CI Docker build (`.github/workflows/test_image.yml`) was failing as it couldn't find the extension to copy. This has been fixed by creating a *dummy* `extensions/bootstrap` file during the CI job. This satisfies the `COPY` command without needing to download the real extension.
*   **.gitignore**: The `extensions/` directory and `extension.zip` file are now ignored to prevent committing downloaded artifacts.

#### 4. Comprehensive Testing

*   New test suites have been added to ensure the reliability of the new features:
    *   `tests/tests_functions/test_secret_rotator.py`: Covers all steps of the secret rotation logic.
    *   `tests/tests_utils/test_secrets.py`: Validates the new shared secrets module, including the extension and boto3 fallback paths.
    *   `tests/tests_views/test_base.py`: Includes new tests for the `/healthz` endpoint.
    *   `tests/tests_utils/test_tokens_manager.py`: Updated to reflect the refactoring to `shared.secrets`.

#### 5. Documentation & Environment Consistency

*   The required Python version for the project has been downgraded to **3.12** across all scripts (`install-dev.sh`, `uninstall-dev.sh`) and documentation (`docs/INSTALL-dev.md`) to support the zappa version.
*   The virtual environment name in the setup documentation has been updated to `env312` to match the `.gitignore` configuration.

## How to Test

1. Run pytest
2. Log in to staging website
3. Go to flask-secret-key-staging in AWS Secrets Manager -> Rotation -> Rotate secret immediately
4. Refresh the staging website. It should take longer to load, as the Lambda function has refreshed, then load properly.

## Checklist

- [x] New code includes a license statement.
- [x] Any TODOs are made into new issues.
- [x] Code is properly linted (for Markdown, TypeScript, and Python).
- [x] Unit tests are written for any new code.
- [x] All existing unit tests pass.
- [x] Any new endpoints include proper AWS Cognito and/or RBAC protections (not applicable for /healthz).

chris01b committed:
Create centralized module for secret retrieval

This commit introduces the core of the new functionality: a robust, centralized module for fetching secrets. It's designed to be environment-aware, using the AWS Parameters and Secrets Lambda Extension when available and gracefully falling back to the boto3 SDK for local development. It also introduces a SecretPayload dataclass for more structured and predictable responses.

chris01b committed:
Integrate centralized secrets module

This commit refactors the entire application to use the new centralized secrets module. By removing direct boto3 calls from various parts of the codebase, we reduce code duplication, improve consistency, and make future updates to our secret-handling strategy much simpler.

chris01b committed:
Add AWS Secrets Extension to Docker builds

This commit updates our infrastructure and deployment configurations. The Dockerfile for both the main application and the wearable_data_retrieval function are modified to include the AWS Parameters and Secrets Lambda Extension. The deployment scripts are enhanced to automate the process of downloading and packaging the extension during the build, except in wearable-data-retrieval.

chris01b committed:
Ensure local development build succeeds

chris01b committed:
Add tests for secrets module and update for refactor

chris01b committed:
Enhance secret handling for rotation

- Updates `shared/secrets.py` to gracefully handle secrets that are raw strings, not just JSON objects. This is a prerequisite for managing the `FLASK_SECRET_KEY` which is a simple string.
- Modifies the `SecretPayload` dataclass to include the raw `secret_string` alongside the parsed `secret_dict`.
- Updates the corresponding unit tests in `tests/tests_utils/test_secrets.py` to align with the new logic and ensure non-JSON secrets are handled correctly.

chris01b committed:
Add four-step secret rotator function

- Implements a new Lambda function in `functions/secret_rotator/` to handle the automatic rotation of the `FLASK_SECRET_KEY`.
- The `handler.py` contains the standard four-step rotation logic required by AWS Secrets Manager: `createSecret`, `setSecret`, `testSecret`, and `finishSecret`.
- Includes a `Dockerfile` and `requirements.txt` to build and package the rotator function with its `requests` dependency.

chris01b committed:
Integrate main application with secret rotation

- Modifies `run.py` to be aware of the rotation process. It now checks for a `SECRET_VERSION_STAGE` environment variable to fetch either the `AWSPENDING` or `AWSCURRENT` version of the `FLASK_SECRET_KEY`. It also stores the secret's version ID in the app config.
- Adds a new, lightweight `/healthz` endpoint to `backend/views/base.py`. This endpoint exposes the currently loaded secret version ID, allowing the rotator's `testSecret` step to verify the correct key is active.

chris01b committed:
Add missing secret loading to RDS stopper

- Updates `rds_stopper.py` to correctly load its configuration from AWS Secrets Manager at runtime.
- Adds the necessary `get_secret` import and invocation logic that was missing.

chris01b committed:
Update deployment scripts for secret rotator

- Updates both `deploy-staging.sh` and `deploy-prod.sh` to manage the deployment of the new secret rotator Lambda function.
- Adds logic to:
  - Build and push the rotator's Docker image to ECR.
  - Deploy or update the rotator Lambda using Zappa.
  - Dynamically create and apply the necessary IAM role policy for the rotator.
  - Set the rotator's environment variables (`APP_LAMBDA_FUNCTION_NAME` and `APP_URL`) after deployment.
- Adds a `--no-rotator` flag to skip this process when desired.

chris01b committed:
Add unit tests for secret rotation feature

- Adds comprehensive unit tests for the new secret rotation functionality.
- Creates `tests/tests_functions/test_secret_rotator.py` to test the rotator Lambda's logic using mock AWS clients.
- Updates `tests/tests_views/test_base.py` to add tests for the new `/healthz` endpoint.

chris01b committed:
Downgrade Python version to 3.12 for zappa deploy requirement

- `install-dev.sh`: Updated the version check to require Python 3.12.
- `uninstall-dev.sh`: Updated the version check to require Python 3.12.
- `docs/INSTALL-dev.md`: Updated the setup guide and commands to reference Python 3.12.

chris01b committed:
Correct exception handling and test logic

1.  In `functions/secret_rotator/handler.py`, the `create_secret` function was attempting to catch a mock exception object (`service_client.exceptions.ResourceNotFoundException`) which does not inherit from `BaseException`, leading to a `TypeError` at runtime. This has been corrected to properly catch `botocore.exceptions.ClientError` and inspect the error code, ensuring robust exception handling.

2.  The `test_finish_secret_updates_version_stage` test in `tests/tests_functions/test_secret_rotator.py` was failing due to an incorrectly configured mock. The `boto3.client` patch has been fixed to correctly provide the `lambda_client` mock, allowing the test to verify the cleanup of environment variables as intended.

Additionally, the `test_secret` function signature has been refactored to remove unused parameters (`service_client`, `arn`), and all corresponding calls have been updated.

chris01b committed:
Address static analysis warnings

- Adds `# noqa: S105` comments to suppress false-positive warnings for hardcoded strings that are used as required AWS API values (e.g., "AWSPENDING") or as identifiers and mock data within the test suite.
- Updates docstrings in test fixtures and helper functions to use the imperative mood, conforming to style guidelines.
- Combines a nested `with` statement in `test_handler_raises_if_rotation_not_enabled` into a single, more readable line.

chris01b committed:
Create dummy Lambda extension file for test_image.yml

chris01b committed:
Add license to handler.py

JStover95 commented on deploy-staging.sh:
```diff
@@ -107,6 +150,99 @@ else
+    ROTATOR_FUNCTION_NAME="${ROTATOR_PROJECT_NAME}-${ROTATOR_STAGE}"
+    ROTATOR_ROLE_NAME="${ROTATOR_FUNCTION_NAME}-ZappaLambdaExecutionRole"
+    SECRET_NAME="flask-secret-key-staging"
+    APP_FUNCTION_NAME="aws-portal-staging-staging"
```
This can be retrieved from zappa instead of hardcoded: `zappa status staging -j | jq '."Lambda Name"'`. I recommend doing this here and in `deploy-prod.sh`.

JStover95 commented on backend/utils/lambda_credentials_manager.py:
```diff
@@ -62,16 +64,7 @@ def get_credentials(self) -> Credentials:
-                logger.error(msg)
-                raise ValueError(msg)
-            secret_data = json.loads(secret_string)
+            secret_data = get_secret(self.secret_name).secret_dict
```
`secret_dict` can be `None`. In this case, and elsewhere where `secret_dict` is accessed, we should check whether it is `None`. If it is, an error should be logger and thrown. If the secret is not required, then a warning should be logged.

JStover95 commented on backend/views/base.py:
```diff
@@ -118,3 +118,21 @@ def touch():
+
+
+@blueprint.route("/healthz")
+def healthz():
```
Can we combine this with our existing health check endpoint? We can discuss via call.

JStover95 commented on functions/secret_rotator/handler.py:
```diff
@@ -0,0 +1,283 @@
+
+    arn = event["SecretId"]
+    token = event["ClientRequestToken"]
+    step = event["Step"]
```
I recommend validating `token` and `step`. We can log and throw an error if they aren't the expected values.

JStover95 replied:
Specifically, just check that `token` is in `versions`.

JStover95 reviewed (changes requested):
Looks great. Only a few minor comments.

One more recommendation is to add `boto3-stubs[cognito-idp, cloudformation, dynamodb, ecr, lambda, logs, rds, secretsmanager, sts]` to our root requirements. It's just handy for type-hinting our different clients.

JStover95 commented on functions/secret_rotator/handler.py:
```diff
@@ -0,0 +1,283 @@
+    # tuning, but a cold start for a simple Flask app on Lambda should be
+    # well under this.
+    logger.info("Waiting for application to restart with new secret...")
+    time.sleep(15)
```
I recommend using `lambda_client.get_function` to wait for the function to be ready. We can check "State" of the response and wait until it is either "Active" or "Failed". See: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/lambda/client/get_function.html

JStover95 commented:
I tried implementing `secrets.py` and came across two more recommendations:

1. I recommend wrapping `get_secret` in a generic `LambdaSecretProvider` and making `SecretPayload` generic. For example:

```python
from typing import TypeVar

T = TypeVar("T")


@dataclass
class SecretPayload[T]:
    """Represents the data retrieved for a secret from AWS Secrets Manager."""

    version_id: str
    secret_string: str | None = None
    secret_dict: T | None = None
    version_stages: list[str] = field(default_factory=list)
    name: str | None = None
    arn: str | None = None


class SecretProvider[T]:
    def __init__(self, secret_name: str):
        self.secret_name = secret_name

    def get_secret(self) -> SecretPayload[T]:
        return get_secret(self.secret_name)
```

Now `secret_dict` will be properly typed. For example:

```python
class AppSyncConfig(TypedDict):
    host: str
    access_key: str
    region: str
    secret_key: str


secret_name = os.getenv("AWS_APP_SYNC_SECRET_NAME")
app_sync_config = SecretProvider[AppSyncConfig](secret_name).get_secret()

# Dictionary keys will now be typed
print(app_sync_config.secret_dict["host"])
print(app_sync_config.secret_dict["region"])
```

2. I also recommend renaming `secrets.py` to something like `lambda_secrets.py` or `lambda_secrets_provider.py`. `secrets` is a builtin Python, and the filename could lead to conflicts.

This will be very helpful for us to include in other repos, so we should consider publishing this to PyPi under an open-source ditti_utils package, but this is a low priority for now.

chris01b committed:
Retrieve APP_FUNCTION_NAME from Zappa instead of hardcoding

chris01b committed:
Add more handling for None secret_dict from get_secret

- Add explicit checks for secret_dict being None after calling get_secret in all locations where secrets are required.
- In LambdaCredentialsManager, raise an error if the secret is missing or empty.
- In wearable_data_retrieval Lambda, raise ConfigFetchError if config or tokens secrets are missing or empty.
- In TokensManager, treat a None secret_dict as an empty dict and log a warning, preserving non-fatal behavior for optional secrets.

chris01b committed:
Update secret rotator to use unified /health endpoint

- Changed the secret rotation Lambda (handler.py) to use /health instead of /healthz for health checks.
- Updated all tests in test_secret_rotator.py to expect /health instead of /healthz.

chris01b committed:
Add boto3-stubs[cognito-idp, cloudformation, dynamodb, ecr, lambda, logs, rds, secretsmanager, sts]==1.34.144 to requirements.txt

chris01b committed:
Poll for application to restart with new secret intead of hard-coded wait

chris01b committed:
Fix linting

chris01b committed:
Add generic SecretProvider and type-safe SecretPayload for AWS secrets

- Make SecretPayload a generic dataclass, allowing secret_dict to be strongly typed.
- Add SecretProvider[T] class for type-safe retrieval of secrets, enabling static type checking and IDE autocompletion for secret schemas.
- Preserve full backward compatibility: existing get_secret usage and imports continue to work.
- Update docstrings and comments for production quality and clarity.

chris01b committed:
Rename secrets.py to lambda_secrets_provider.py and update uses

chris01b committed:
Fix npm package vulnerability

JStover95 reviewed (approved):
Looks good!
